File: Applied Course/4.Problem Solving/5.Problems on Trees/35.Kth Smallest Element in a BST.cpp
#include<bits/stdc++.h>
using namespace std;
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
// Logic-1
class Solution {
public:
    vector<int> v;
    vector<int> inorder(TreeNode* root)
    {
        if(!root)
            return v;
        inorder(root->left);
        v.push_back(root->val);
        inorder(root->right);
        return v;
    }
    int kthSmallest(TreeNode* root, int k) {
        return inorder(root)[k-1];
    }
};
// Logic-2
class Solution {
public:
    int kthSmallest(TreeNode* root, int k) {
        stack<TreeNode*> s;
        while(true)
        {
            while(root)
            {
                s.push(root);
                root=root->left;
            }
            root=s.top();
            s.pop();
            k-=1;
            if(k==0)
                return root->val;
            root=root->right;
        }
    }
};
int main()
{
    // https://leetcode.com/problems/kth-smallest-element-in-a-bst/
    return 0;
}

File: Applied Course/4.Problem Solving/5.Problems on Trees/35.Kth Smallest Element in a BST.py
# https://leetcode.com/problems/kth-smallest-element-in-a-bst/
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
# Logic-1
class Solution:
    def kthSmallest(self, root: TreeNode, k: int) -> int:
        def inorder(root):
            return inorder(root.left)+[root.val]+inorder(root.right) if root else []
        lst=inorder(root)
        return lst[k-1]
# Logic-2
class Solution:
    def kthSmallest(self, root: TreeNode, k: int) -> int:
        stack=[]
        while True:
            while root:
                stack.append(root)
                root=root.left
            root=stack.pop()
            k-=1
            if k==0:
                return root.val
            root=root.right